pjmedia/src/pjmedia/sdp_neg.c

```diff
@@ -1793,3 +1946,167 @@ PJ_DEF(pj_status_t) pjmedia_sdp_neg_fmt_match(pj_pool_t *pool,
                             offer, o_fmt_idx, answer, a_fmt_idx, option);
 }
 
+static unsigned assign_new_pt(pj_pool_t *pool, media_pt_map *used_pt, 
+                              const pj_str_t *codec)
+{
+    unsigned pt_check;
+    unsigned i = 0;
+    unsigned used_pt_num = used_pt->map_cnt;
+    unsigned max_pt_cnt = PJ_ARRAY_SIZE(used_pt->map);
+    const pj_str_t telephone_event = pj_str("telephone-event");
+
+    if (pj_strncmp(codec, &telephone_event, telephone_event.slen) == 0)
+        pt_check = PJMEDIA_RTP_PT_TELEPHONE_EVENTS;
+    else
+        pt_check = START_DYNAMIC_PT;
+    
+    for (; i < used_pt_num && i < max_pt_cnt; ++i) {
+        if (pj_strcmp(codec, &used_pt->map[i].codec) == 0) {
+            return used_pt->map[i].pt;
+        }
+    }
+    i = 0;
+
+    while ((i < used_pt_num) && (i < max_pt_cnt) && (pt_check <= 127))
+    {
+        if (pt_check == used_pt->map[i].pt) {
+            pt_check++;
+            i = 0;
+        } else {
+            i++;
+        }
+    }
+    if (pt_check > 127 || i == max_pt_cnt) {
+        /* No more available PT */
+        return 0;
+    }
+
+    used_pt->map[used_pt->map_cnt].pt = pt_check;
+    pj_strdup(pool, &used_pt->map[used_pt->map_cnt++].codec, codec);
+
+    return pt_check;
+}
+
+PJ_DEF(pj_status_t) pjmedia_sdp_neg_validate_pt(pjmedia_sdp_neg* neg,
+                                                pjmedia_sdp_session* sdp_sess)
+{
+#if PJMEDIA_SDP_NEG_MAINTAIN_SESSION_PT
+    pj_status_t status = PJ_SUCCESS;
+    unsigned i = 0;
+
+    for (; i < sdp_sess->media_count && i < neg->med_pt_map_cnt; ++i) {
+        unsigned j = 0;
+        unsigned attr_cnt = 0;
+        pjmedia_sdp_attr* new_attr_list[PJMEDIA_MAX_SDP_ATTR];
+        pjmedia_sdp_attr* ori_attr_list[PJMEDIA_MAX_SDP_ATTR];
+        pj_str_t ori_fmt[PJMEDIA_MAX_SDP_FMT];
+        pjmedia_sdp_media* sess_media = sdp_sess->media[i];
+        media_pt_map used_pt_map;
+        media_pt_map* med_pt_map = get_media_pt_map(neg, &sdp_sess->media[i]->desc.media);
+
+        if (!med_pt_map) {
+            PJ_LOG(4, (THIS_FILE, "Unable to find media pt map for [%.*]",
+                       sdp_sess->media[i]->desc.media.slen, 
+                       sdp_sess->media[i]->desc.media.ptr));
```

Comment:
To help troubleshooting, for example reading this line in a log file in the future when we may already forget this feature exists, there will be questions like "who & why find pt map?", "what happen if it is not found", so perhaps better add something like "..., PTs will not be validated for this media" for initial hints on reading logs?